src/atc/delta/delta_handle.py

```diff
@@ -99,6 +103,88 @@ def recreate_hive_table(self):
         self.drop()
         self.create_hive_table()
 
-    def merge(self):
-        # This should be a method which can merge from delta to delta
-        raise NotImplementedError
+    def get_tablename(self) -> str:
+        return self._name
+
+    def upsert(
+        self,
+        df: DataFrame,
+        join_cols: List[str],
+        incremental_load: bool = True,
+    ) -> Union[DataFrame, None]:
+        def do_simple_save(
+            df_source: DataFrame,
+            incremental_load_local: bool = True,
+        ):
+            mode = "append" if incremental_load_local else "overwrite"
+
+            self.write_or_append(df_source, mode=mode)
+            print(
+                "Incremental Base - incremental load with append"
+                if incremental_load_local
+                else "Incremental Base - full load with append"
+            )
+            return df_source
+
+        if df is None:
+            return None
+
+        # check null keys in our dataframe.
+        any_null_keys = len(
+            df.filter(" OR ".join(f"({col} is NULL)" for col in join_cols)).take(1)
+        )
+
+        if any_null_keys:
+            print(
+                "Null keys found in input dataframe. "
+                "Rows will be discarded before load."
+            )
+            df = df.filter(" AND ".join(f"({col} is NOT NULL)" for col in join_cols))
+
+        # Load data from the target table for the purpose of incremental load
+        if not incremental_load:
+            return do_simple_save(df, incremental_load_local=False)
+
+        df_target = self.read()
+
+        # If the target is empty, always do faster full load
+        if len(df_target.take(1)) == 0:
```

Comment:
This staement of full load when target is empty. Could be hanndled by the merge_required.

Reply:
It could. But this will force an overwrite, where CheckDfMerge forces an append. I will keep them splittet for avoiding too much complexity.